sycl/source/detail/kernel_bundle_impl.hpp

```diff
@@ -263,97 +330,117 @@ class kernel_bundle_impl
+                AOTDevImgWithDeps, bundle_state::executable);
+          }
+
+          // Record all the AOT exported symbols and kernels.
+          std::unordered_set<std::string_view> AOTExportedSymbols;
+          std::unordered_set<std::string_view> AOTKernelNames;
+          for (const RTDeviceBinaryImage *AOTBin : AOTBinaries) {
+            for (const sycl_device_binary_property &ESProp :
+                 AOTBin->getExportedSymbols())
+              AOTExportedSymbols.insert(ESProp->Name);
+            for (const sycl_device_binary_property &KNProp :
+                 AOTBin->getKernelNames())
+              AOTKernelNames.insert(KNProp->Name);
+          }
+
+          for (device_image_impl &DevImg : ObjectBundleImpl->device_images()) {
+            // If the image is the same as one of the offline images, we can
+            // skip it.
+            if (OfflineDeviceImageSet.find(&DevImg) !=
+                OfflineDeviceImageSet.end())
+              continue;
+
+            // If any of the exported symbols overlap with an AOT binary, skip
+            // this image.
+            for (const sycl_device_binary_property &ESProp :
+                 DevImg.get_bin_image_ref()->getExportedSymbols())
+              if (AOTExportedSymbols.find(ESProp->Name) !=
+                  AOTExportedSymbols.end())
+                continue;
+
+            // If any of the kernels overlap with an AOT binary, skip this
+            // image.
+            for (const sycl_device_binary_property &KNProp :
+                 DevImg.get_bin_image_ref()->getKernelNames())
+              if (AOTKernelNames.find(KNProp->Name) != AOTKernelNames.end())
+                continue;
+
             DevImagesSet.insert(&DevImg);
-      DevImages.reserve(DevImagesSet.size());
-      for (auto It = DevImagesSet.begin(); It != DevImagesSet.end();)
-        DevImages.push_back(createSyclObjFromImpl<device_image_plain>(
-            *DevImagesSet.extract(It++).value()));
-    }
+          }
+        }
 
-    // Check for conflicting kernels in RTC kernel bundles.
-    {
-      std::set<std::string_view, std::less<>> SeenKernelNames;
-      std::set<std::string_view, std::less<>> Conflicts;
-      for (const device_image_plain &DevImage : DevImages) {
-        const KernelNameSetT &KernelNames =
-            getSyclObjImpl(DevImage)->getKernelNames();
-        std::vector<std::string_view> Intersect;
-        std::set_intersection(SeenKernelNames.begin(), SeenKernelNames.end(),
-                              KernelNames.begin(), KernelNames.end(),
-                              std::inserter(Conflicts, Conflicts.begin()));
-        SeenKernelNames.insert(KernelNames.begin(), KernelNames.end());
-      }
+        DevImages.reserve(DevImages.size() + DevImagesSet.size());
+        for (auto It = DevImagesSet.begin(); It != DevImagesSet.end();)
+          DevImages.push_back(createSyclObjFromImpl<device_image_plain>(
+              *DevImagesSet.extract(It++).value()));
```

Comment:
Can this be changed to `std::move_iterator` instead? It feels like there is some magic happening here, but it's not clear what's the intent exactly.